cmd/syncComparator.go

```diff
@@ -54,30 +199,192 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+					// We should not find this entry as it should be removed as part of parent enumeration only exception to this is root folder (that's what we checking in if statement).
+					// But due to go routine scheduling it may happen before parent enumeration at target finish or even start, other goroutine picked the child entry and finish the enumeration.
+					// In that case we don't need to do anything. If folder is empty and don't exist then will be created as parent enumeration done.
+					// If folder is not empty then it will be created by it's children. Only thing, it's metadata will be set by parent only as it has the required info.
+					fmt.Printf("lcRelativePath[%s], We should not find this entry as it should be removed as part of parent enumeration."+
+						"\n But due to go routine scheduling it may happen before parent enumeration at target finish child done with enumeration."+
+						"\n If folder is empty and don't exist then will be created as parent enumeration done. Otherwise will be created by its child enumeration."+
+						"\n Its metadata will be set by parent only as it has the required info.", lcRelativePath)
+				}
+			}
+
+			lcFolderName = path.Join(lcFolderName, lcFileName)
+			foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcFolderName]
+			if folderPresent {
+				f.FinalizeTargetDirectory(foldermap, destinationObject.isFinalizeAll)
+			}
+			return nil
+		} else {
+			if folderPresent {
+				sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+				// Lets check if entry in source is folder and entry on destination should be folder.
+				// Otherwise folder in source changed to file.
+				if present && (destinationObject.entityType == sourceObjectInMap.entityType) {
+					delete(foldermap.indexMap, lcFileName)
+					size := storedObjectSize(sourceObjectInMap)
+					size = -size
+					atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
+					if sourceObjectInMap.relativePath != destinationObject.relativePath {
+						panic("Relative Path not matched")
+					}
+					delete(f.sourceFolderIndex.folderMap[lcFolderName].indexMap, lcFileName)
+
+					// In case of folder if ctime/mtime changed, lets create the folder.
+					dataChanged, metadataChanged := f.HasFileChangedSinceLastSyncUsingTargetCompare(destinationObject, sourceObjectInMap)
+					if f.disableComparison || dataChanged || metadataChanged {
+						err := f.copyTransferScheduler(sourceObjectInMap)
+						if err != nil {
+							return err
+						}
+					}
+					return nil
+				}
+			}
+			// We detect folder not present on source, now we need to delete the folder and files underneath.
+			// Other case will be folder at source changed to file, so we need to delete the folder in target.
+			// TODO: Need to add call to delete the folder.
+			_ = f.destinationCleaner(destinationObject)
+
+			return nil
+		}
+	}
+
+	// File case.
+	// Parent Folder present, we need to check if file exists or not.
+	if folderPresent {
```

Comment:
again, we should never have the case that folderPresent  is false. we can panic on this,